[CI] LatLonGeometryRelationVisitorTests testPoint failing

May be related to #93889

**Build scan:**
https://gradle-enterprise.elastic.co/s/necv7blkxj2ak/tests/:x-pack:plugin:spatial:test/org.elasticsearch.xpack.spatial.index.fielddata.LatLonGeometryRelationVisitorTests/testPoint

**Reproduction line:**
```
./gradlew ':x-pack:plugin:spatial:test' --tests "org.elasticsearch.xpack.spatial.index.fielddata.LatLonGeometryRelationVisitorTests.testPoint" -Dtests.seed=E85F5ED917D81660 -Dtests.locale=ro -Dtests.timezone=America/St_Vincent -Druntime.java=20
```

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.spatial.index.fielddata.LatLonGeometryRelationVisitorTests&tests.test=testPoint

**Failure excerpt:**
```
java.lang.AssertionError: WITHIN/QUERY_INSIDE: Geometry Point(0.0,0.0) relates to shape MultiPoint: QUERY_INSIDE
Expected: <false>
     but: was <true>

  at __randomizedtesting.SeedInfo.seed([E85F5ED917D81660:D5007492732D0B33]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.elasticsearch.xpack.spatial.index.fielddata.LatLonGeometryRelationVisitorTests.doTestShape(LatLonGeometryRelationVisitorTests.java:206)
  at org.elasticsearch.xpack.spatial.index.fielddata.LatLonGeometryRelationVisitorTests.doTestShape(LatLonGeometryRelationVisitorTests.java:159)
  at org.elasticsearch.xpack.spatial.index.fielddata.LatLonGeometryRelationVisitorTests.doTestShape(LatLonGeometryRelationVisitorTests.java:154)
  at org.elasticsearch.xpack.spatial.index.fielddata.LatLonGeometryRelationVisitorTests.doTestShapes(LatLonGeometryRelationVisitorTests.java:144)
  at org.elasticsearch.xpack.spatial.index.fielddata.LatLonGeometryRelationVisitorTests.testPoint(LatLonGeometryRelationVisitorTests.java:41)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  at java.lang.reflect.Method.invoke(Method.java:578)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1623)

```

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/fielddata/LatLonGeometryRelationVisitorTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/fielddata/LatLonGeometryRelationVisitorTests.java
@@ -17,9 +17,15 @@ import org.apache.lucene.tests.geo.GeoTestUtil;
 import org.elasticsearch.common.geo.GeometryNormalizer;
 import org.elasticsearch.common.geo.Orientation;
 import org.elasticsearch.geo.GeometryTestUtils;
+import org.elasticsearch.geometry.Circle;
 import org.elasticsearch.geometry.Geometry;
+import org.elasticsearch.geometry.GeometryCollection;
+import org.elasticsearch.geometry.GeometryVisitor;
 import org.elasticsearch.geometry.LinearRing;
+import org.elasticsearch.geometry.MultiLine;
 import org.elasticsearch.geometry.MultiPoint;
+import org.elasticsearch.geometry.MultiPolygon;
+import org.elasticsearch.geometry.Rectangle;
 import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.xpack.spatial.util.GeoTestUtils;
 
@@ -135,6 +141,18 @@ public class LatLonGeometryRelationVisitorTests extends ESTestCase {
         }
     }
 
+    /** Explicitly test failure found in <a href="https://github.com/elastic/elasticsearch/issues/98063">#98063</a> */
+    public void testOriginPointInMultipoint() throws Exception {
+        ArrayList<org.elasticsearch.geometry.Point> points = new ArrayList<>();
+        points.add(new org.elasticsearch.geometry.Point(0.0, 0.0));
+        points.add(new org.elasticsearch.geometry.Point(0.0, 0.0));
+        points.add(new org.elasticsearch.geometry.Point(0.0, 1.401298464324817E-45));
+        Geometry geometry = new MultiPoint(points);
+        GeoShapeValues.GeoShapeValue geoShapeValue = GeoTestUtils.geoShapeValue(geometry);
+        GeometryDocValueReader reader = GeoTestUtils.geometryDocValueReader(geometry, CoordinateEncoder.GEO);
+        doTestShape(geometry, geoShapeValue, reader, new Point(0, 0));
+    }
+
     private <T extends LatLonGeometry> void doTestShapes(Supplier<T> supplier) throws Exception {
         Geometry geometry = GeometryNormalizer.apply(Orientation.CCW, GeometryTestUtils.randomGeometryWithoutCircle(0, false));
         GeoShapeValues.GeoShapeValue geoShapeValue = GeoTestUtils.geoShapeValue(geometry);
@@ -161,18 +179,7 @@ public class LatLonGeometryRelationVisitorTests extends ESTestCase {
 
     private boolean isIdenticalPoint(Geometry geometry, LatLonGeometry latLonGeometry) {
         if (latLonGeometry instanceof Point latLonPoint) {
-            if (geometry instanceof org.elasticsearch.geometry.Point point) {
-                return encodeLatitude(point.getLat()) == encodeLatitude(latLonPoint.getLat())
-                    && encodeLongitude(point.getLon()) == encodeLongitude(latLonPoint.getLon());
-            } else if (geometry instanceof org.elasticsearch.geometry.Line line) {
-                for (int i = 0; i < line.length(); i++) {
-                    if (encodeLatitude(line.getLat(i)) != encodeLatitude(latLonPoint.getLat())
-                        || encodeLongitude(line.getLon(i)) != encodeLongitude(latLonPoint.getLon())) {
-                        return false;
-                    }
-                }
-                return true;
-            }
+            return geometry.visit(new TestIdenticalPointVisitor(latLonPoint));
         }
         return false;
     }
@@ -279,5 +286,114 @@ public class LatLonGeometryRelationVisitorTests extends ESTestCase {
 
     private double quantizeLon(double lon) {
         return decodeLongitude(encodeLongitude(lon));
+    }
+
+    /**
+     * This visitor returns false if any point in the geometry is not identical to the provided point.
+     * Identical means that the encoded lat and lon values are the same.
+     */
+    private static class TestIdenticalPointVisitor implements GeometryVisitor<Boolean, RuntimeException> {
+        private final int encodedLat;
+        private final int encodedLon;
+
+        private TestIdenticalPointVisitor(Point latLonPoint) {
+            encodedLat = encodeLatitude(latLonPoint.getLat());
+            encodedLon = encodeLongitude(latLonPoint.getLon());
+        }
+
+        private boolean isIdenticalPoint(double lat, double lon) {
+            return encodeLatitude(lat) == encodedLat && encodeLongitude(lon) == encodedLon;
+        }
+
+        @Override
+        public Boolean visit(Circle circle) {
+            if (circle.getRadiusMeters() == 0) {
+                return isIdenticalPoint(circle.getLat(), circle.getLon());
+            }
+            return false;
+        }
+
+        @Override
+        public Boolean visit(GeometryCollection<?> collection) {
+            for (Geometry shape : collection) {
+                if (shape.visit(this) == false) {
+                    return false;
+                }
+            }
+            return collection.size() > 0;
+        }
+
+        @Override
+        public Boolean visit(org.elasticsearch.geometry.Line line) {
+            for (int i = 0; i < line.length(); i++) {
+                if (isIdenticalPoint(line.getLat(i), line.getLon(i)) == false) {
+                    return false;
+                }
+            }
+            return line.length() > 0;
+        }
+
+        @Override
+        public Boolean visit(LinearRing ring) {
+            return visit((org.elasticsearch.geometry.Line) ring);
+        }
+
+        @Override
+        public Boolean visit(MultiLine multiLine) {
+            for (org.elasticsearch.geometry.Line line : multiLine) {
+                if (visit(line) == false) {
+                    return false;
+                }
+            }
+            return multiLine.size() > 0;
+        }
+
+        @Override
+        public Boolean visit(MultiPoint multiPoint) {
+            for (org.elasticsearch.geometry.Point point : multiPoint) {
+                if (visit(point) == false) {
+                    return false;
+                }
+            }
+            return multiPoint.size() > 0;
+        }
+
+        @Override
+        public Boolean visit(MultiPolygon multiPolygon) {
+            for (org.elasticsearch.geometry.Polygon polygon : multiPolygon) {
+                if (visit(polygon) == false) {
+                    return false;
+                }
+            }
+            return multiPolygon.size() > 0;
+        }
+
+        @Override
+        public Boolean visit(org.elasticsearch.geometry.Point point) {
+            return isIdenticalPoint(point.getLat(), point.getLon());
+        }
+
+        @Override
+        public Boolean visit(org.elasticsearch.geometry.Polygon polygon) {
+            if (visit(polygon.getPolygon()) == false) {
+                return false;
+            }
+            for (int i = 0; i < polygon.getNumberOfHoles(); i++) {
+                LinearRing hole = polygon.getHole(i);
+                if (visit(hole) == false) {
+                    return false;
+                }
+            }
+            return polygon.getPolygon().length() > 0;
+        }
+
+        @Override
+        public Boolean visit(Rectangle rectangle) {
+            int eMinX = encodeLongitude(rectangle.getMinX());
+            int eMaxX = encodeLongitude(rectangle.getMaxX());
+            int eMinY = encodeLatitude(rectangle.getMinY());
+            int eMaxY = encodeLatitude(rectangle.getMaxY());
+            return eMinX == eMaxX && eMinY == eMaxY && isIdenticalPoint(rectangle.getMinLat(), rectangle.getMinLon());
+        }
     }
 }

